internal/api/proxy.go

```diff
@@ -0,0 +1,139 @@
+package api
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+
+	"golang.org/x/net/proxy"
+)
+
+func applyProxy(transport *http.Transport, proxyEndpointURL *url.URL, proxyEndpointPath string) (applied bool) {
+	if proxyEndpointURL == nil && proxyEndpointPath == "" {
+		return false
+	}
+
+	handshakeTLS := func(conn net.Conn, addr string) (net.Conn, error) {
+		// Extract the hostname (without the port) for TLS SNI
+		host, _, err := net.SplitHostPort(addr)
+		if err != nil {
+			return nil, err
+		}
+		// Be sure to clone TLS-specific config settings from the transport,
+		// like InsecureSkipVerify.
+		tlsConn := tls.Client(conn, &tls.Config{
+			ServerName:         host,
+			InsecureSkipVerify: transport.TLSClientConfig.InsecureSkipVerify,
+		})
+		if err := tlsConn.Handshake(); err != nil {
+			return nil, err
+		}
+		return tlsConn, nil
+	}
+
+	var dial func(ctx context.Context, network, addr string) (net.Conn, error)
+	var dialTLS func(ctx context.Context, network, addr string) (net.Conn, error)
+
+	if proxyEndpointPath != "" {
+		dial = func(_ context.Context, _, _ string) (net.Conn, error) {
+			return net.Dial("unix", proxyEndpointPath)
+		}
+		dialTLS = func(ctx context.Context, network, addr string) (net.Conn, error) {
+			conn, err := dial(ctx, network, addr)
+			if err != nil {
+				return nil, err
+			}
+			return handshakeTLS(conn, addr)
+		}
+	} else if proxyEndpointURL != nil {
+		if proxyEndpointURL.Scheme == "socks5" ||
+			proxyEndpointURL.Scheme == "socks5h" {
+			dial = func(_ context.Context, network, addr string) (net.Conn, error) {
+				// figure out the proxy every dial because we have error handling here.
+				// In NewClient, we don't have error handling; all we can do there is panic.
+
+				// FromURL really only handles SOCKS5 (unless other schemes have ben registered),
+				// but since it also handles credentials,
+				// we'll use it instead of manually handling any credentials embedded in the URL,
+				// and calling SOCKS5 ourself.
+				dialer, err := proxy.FromURL(proxyEndpointURL, proxy.Direct)
+				if err != nil {
+					return nil, err
+				}
+				return dialer.Dial(network, addr)
+			}
+			dialTLS = func(ctx context.Context, network, addr string) (net.Conn, error) {
+				// Dial the underlying connection through the proxy
+				conn, err := dial(ctx, network, addr)
+				if err != nil {
+					return nil, err
+				}
+				return handshakeTLS(conn, addr)
+			}
+		} else if proxyEndpointURL.Scheme == "http" || proxyEndpointURL.Scheme == "https" {
```

Comment:
Why are we manually implementing this? It looks to me like is all supported by the standard library by setting `(*http.Transport).Proxy`. 

In fact, I think everything but sockets is likely already supported with no code changes by setting `HTTP_PROXY` and `HTTPS_PROXY`. Take a look at `DefaultTransport`, which defaults to respecting [httpproxy.FromEnvironment](https://pkg.go.dev/golang.org/x/net/http/httpproxy#FromEnvironment).

Reply:
Prompted by your question, I tested socks proxies and they do work out of the box.

I hadn't tested them - I started with testing https proxy, which didn't work, so I just dove into implementing them all.

The error I get when using https proxies OOTB (I've used `mitmproxy` locally and https://devproxy.sgdev.org/) is:
```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
https_proxy=https://localhost:8080 \
src login --insecure-skip-verify=true 
❌ Problem: Error communicating with https://sourcegraph.sourcegraph.com: Post "https://sourcegraph.sourcegraph.com/.api/graphql": malformed HTTP status code "protocol"
```
When using `devproxy.sgdev.org`, I see the error `proxyconnect tcp: tls: first record does not look like a TLS handshake`.

When using `mitmproxy` locally, there's a corresponding error `HTTP/2 protocol error: Invalid HTTP/2 preamble.` in its output.

Note that this is the current release of `src` that I'm using here, not a local build (although I get the same behavior with a local build).

Manually dialing and doing the CONNECT for the https proxy fixes the protocol errors. There might be a cleaner way to fix them.

Reply:
Hmmm, legit point about `--insecure-skip-verify`, but I am not able to fix the errors the way you describe, and I get similar errors even when I use a proxy that doesn't require `--insecure-skip-verify`, like https://devproxy.sgdev.org:3128, which has a public CA cert, or `tinyproxy`, which is a tunneling proxy that does not require its own CA cert:
```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
https_proxy=https://devproxy.sgdev.org:3128 \
~/go/bin/src login

❌ Problem: Error communicating with https://sourcegraph.sourcegraph.com: Post "https://sourcegraph.sourcegraph.com/.api/graphql": proxyconnect tcp: tls: first record does not look like a TLS handshake
```

```
SRC_ENDPOINT=https://sourcegraph.sourcegraph.com \
SRC_ACCESS_TOKEN=deadbeef \
https_proxy=https://localhost:8888 \         
~/go/bin/src login

❌ Problem: Error communicating with https://sourcegraph.sourcegraph.com: Post "https://sourcegraph.sourcegraph.com/.api/graphql": proxyconnect tcp: net/http: TLS handshake timeout
```
The connection with `tinyproxy` actually times out.

All of these errors are solved  when I use manual dialing, by changing from `https_proxy` to `SRC_PROXY_ENDPOINT`.

I don't want to be pedantic about this, but I am interested in understanding the actual behavior and resolution, if possible.

Reply:
> I don't want to be pedantic about this

Proxy behavior is exactly the kind of thing that is worth being pedantic about IMO 😄

Reply:
Fair point. :-)

I tried importing the cert into the login keychain (and marking it as "Always trust"), and while that affects the behavior when using `SRC_PROXY_ENDPOINT`, it does not when using `https_proxy`.

Reply:
I'm able to get `http://devproxy.sgdev.org:3128` working, but both `src` and `curl` fail when trying to use `https://devproxy.sgdev.org:3129` for me with an "expired certificate" error

Reply:
Looks to me like that's correct, and that the cert is, in fact, expired. Do you know who runs the devproxy?

Reply:
I confirmed that `src` works against `https://devproxy.sgdev.org:3129` if I set `--insecure-skip-verify` (with the fix for the insecure transport), as does `curl` if I also set `--proxy-insecure`

Reply:
You were turning off HTTP/2 by using `GODEBUG=http2client=0`. The upshot of our long discussion and experimentation about this is that we're manually dialing in order to use the HTTP/1 protocol instead of HTTP/2.

I added a bunch of comments to the code to help our future selves remember this.